library/ssl_cli.c

```diff
@@ -1385,99 +1614,199 @@ static int ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
 #endif /* MBEDTLS_SSL_ALPN */
 
 /*
- * Parse HelloVerifyRequest.  Only called after verifying the HS type.
+ *
+ * STATE HANDLING: ServerHello
+ *
+ */
+
+/*
+ * Overview
  */
+
+/* Main entry point; orchestrates the other functions */
+static int ssl_process_server_hello( mbedtls_ssl_context *ssl );
+
+/* Fetch and preprocess
+ * Returns a negative value on failure, and otherwise
+ * - SSL_SERVER_HELLO_COORDINATE_HELLO or
+ * - SSL_SERVER_HELLO_COORDINATE_VERIFY
+ * to indicate which message is expected and to be parsed next. */
+#define SSL_SERVER_HELLO_COORDINATE_HELLO  0
+#define SSL_SERVER_HELLO_COORDINATE_VERIFY 1
+static int ssl_process_server_hello_coordinate( mbedtls_ssl_context *ssl );
+
+/* Parse ServerHello */
+static int ssl_process_server_hello_parse( mbedtls_ssl_context *ssl,
+                                           const unsigned char* buf,
+                                           size_t buflen );
+static int ssl_process_server_hello_postprocess( mbedtls_ssl_context *ssl );
+/* Parse HelloVerify */
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
-static int ssl_parse_hello_verify_request( mbedtls_ssl_context *ssl )
+static int ssl_process_hello_verify_parse( mbedtls_ssl_context *ssl,
+                                           const unsigned char* buf,
+                                           size_t buflen );
+static int ssl_process_hello_verify_postprocess( mbedtls_ssl_context *ssl );
+#endif
+
+/*
+ * Implementation
+ */
+
+static int ssl_process_server_hello( mbedtls_ssl_context *ssl )
 {
-    const unsigned char *p = ssl->in_msg + mbedtls_ssl_hs_hdr_len( ssl );
-    int major_ver, minor_ver;
-    unsigned char cookie_len;
+    int ret = 0;
+    int msg_expect;
```

Comment:
Maybe msg_expected?

Reply:
I slightly prefer `msg_expect` because what message we expect is determined at the time this function runs, and not beforehand.